Restructure project, add build scripts

iMrShadow opened the pull request:
- Restructured the project.
   - `Editor` and `TelltaleToolLib` are two separate projects under one folder.
   - All libraries and dependencies will be located in the `ThirdParty` folder.
- Added build scripts.
- Removed project specific files.

Tested on:
- [x] Windows - Visual Studio Community 2022
- [x] MacOS - Xcode
- [x] Linux

iMrShadow committed:
Restructure project

iMrShadow committed:
Add build scripts, ignore build folder

iMrShadow committed:
Rename folders

Editor -> TelltaleEditor
TelltaleToolLib -> ToolLibrary

asilz commented on TelltaleEditor/Source/Editor/CMakeLists.txt:
```diff
@@ -0,0 +1,19 @@
+set_target_properties(${TARGET_NAME} PROPERTIES CXX_STANDARD 17 OUTPUT_NAME "TelltaleEditor")
+set_target_properties(${TARGET_NAME} PROPERTIES FOLDER "Editor")
+
+target_compile_definitions(Editor PRIVATE
```
Is release always defined. You can add a check like so perhaps
```
if(CMAKE_BUILD_TYPE STREQUAL "Debug") # Define DEBUG and RELEASE macros depending on cmake build type
    target_compile_definitions(${module} PRIVATE DEBUG=1)
else()
    target_compile_definitions(${module} PRIVATE RELEASE=1)
endif()
```

iMrShadow replied:
Yeah, I did it on purpose, because CMake complained about the macros.

iMrShadow replied:
Fixed.

asilz commented on TelltaleEditor/Source/TelltaleToolLib/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+
+# Create a library from the sources
+add_library(TelltaleToolLib STATIC
+    ${TELLTALE_LIB_HEADERS_DIR}/Meta/Meta.hpp
```
This is fine for now, but we should split into different CMakeLists.txt in the future. This will also allow us to have CMakeLists.txt files for different platforms that we can add depending on the platform

iMrShadow replied:
I agree, but I do now know how the library will look like. I will try to add CMakeLists fom each folder

asilz commented on TelltaleEditor/Source/TelltaleToolLib/CMakeLists.txt:
```diff
@@ -0,0 +1,22 @@
+    ${TELLTALE_LIB_SOURCES_DIR}/Platform/Win64.cpp
+    ${TELLTALE_LIB_SOURCES_DIR}/Scheduler/JobScheduler.cpp
+)
+
```
We need to add a definition for the macros for the different platforms. I can't build otherwise. Something like this might work
```
if(UNIX)
    if(Apple)
        target_compile_definitions(ToolLibrary PRIVATE MACOS=1)
    else()
        target_compile_definitions(ToolLibrary PRIVATE LINUX=1)
    endif()
elseif(WIN32)
    target_compile_definitions(ToolLibrary PRIVATE _WIN64=1)
endif()
```

iMrShadow replied:
Will do.

iMrShadow replied:
Fixed.

asilz reviewed (commented):
Looks really nice. Big improvement to the structure of the project.

iMrShadow committed:
Added platform-specific and build types macros

iMrShadow committed:
Add some compile options

asilz commented on TelltaleEditor/Source/ToolLibrary/CMakeLists.txt:
```diff
@@ -13,10 +13,23 @@ add_library(ToolLibrary STATIC
+# Enable platform specific defines
+if(UNIX)
+    if(Apple)
+        target_compile_definitions(ToolLibrary PRIVATE MACOS=1)
```
These have to be PUBLIC or they have to be defined in the Editor too. My last comment was wrong, sorry about that.

iMrShadow replied:
Fixed. I added it to the Editor for encapsulation reasons.

asilz commented on TelltaleEditor/Source/Editor/CMakeLists.txt:
```diff
@@ -14,6 +13,8 @@ target_link_libraries(${TARGET_NAME} PUBLIC ToolLibrary)
-    RELEASE
-)
+if(CMAKE_BUILD_TYPE STREQUAL "Debug") # Define DEBUG and RELEASE macros depending on cmake build type 
+    target_compile_definitions(Editor PRIVATE DEBUG=1)
```
You could remove this and set the corresponding check in the ToolLibrary to PUBLIC. Not sure what is better.

iMrShadow committed:
Add functions for compiling macros

- Commented out the optimization
- Used the new functions for Editor and ToolLib

iMrShadow committed:
Formatted some code, renamed some variables

iMrShadow committed:
Enable the previously disabled compiling options

iMrShadow committed:
Fix MacOS builds

iMrShadow committed:
Add compilation flags for MacOS builds

LucasSaragosa reviewed (approved):
'APPROVE' performed by Xcode.

LucasSaragosa commented:
looks great

iMrShadow committed:
Add missing headers
